pkg/mongodb/MongodbConsumer.php

```diff
@@ -0,0 +1,181 @@
+    }
+
+    /**
+     * Get polling interval in milliseconds.
+     *
+     * @return int
+     */
+    public function getPollingInterval()
+    {
+        return (int) $this->pollingInterval / 1000;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @return MongodbDestination
+     */
+    public function getQueue()
+    {
+        return $this->queue;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @return MongodbMessage|null
+     */
+    public function receive($timeout = 0)
+    {
+        $timeout /= 1000;
+        $startAt = microtime(true);
+
+        while (true) {
+            $message = $this->receiveMessage();
+
+            if ($message) {
+                return $message;
+            }
+
+            if ($timeout && (microtime(true) - $startAt) >= $timeout) {
+                return;
+            }
+
+            usleep($this->pollingInterval);
+
+            if ($timeout && (microtime(true) - $startAt) >= $timeout) {
+                return;
+            }
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @return MongodbMessage|null
+     */
+    public function receiveNoWait()
+    {
+        return $this->receiveMessage();
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @param MongodbMessage $message
+     */
+    public function acknowledge(PsrMessage $message)
+    {
+        // does nothing
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @param MongodbMessage $message
+     */
+    public function reject(PsrMessage $message, $requeue = false)
+    {
+        InvalidMessageException::assertMessageInstanceOf($message, MongodbMessage::class);
+
+        if ($requeue) {
+            $this->context->createProducer()->send($this->queue, $message);
+
+            return;
+        }
+    }
+
+    /**
+     * @return MongodbMessage|null
+     */
+    protected function receiveMessage()
+    {
+        try {
+            $now = time();
+            $collection = $this->context->getCollection();
+            $message = $collection->findOne(['$or' => [['delayed_until' => ['$exists' => false]], ['delayed_until' => ['$lte' => $now]]]], ['sort' => ['priority' => -1]]);
+            if (!$message) {
+                return null;
+            }
+            $mongodbMessage = $message->getArrayCopy();
+            $convertedMessage = $this->convertMessage($mongodbMessage);
+            $affected = $collection->deleteOne(['_id' => $mongodbMessage['_id']]);
+            if (1 !== $affected->getDeletedCount()) {
+                throw new \LogicException(sprintf('Expected record was removed but it is not. id: "%s"', $mongodbMessage['_id']->__toString()));
+            }
+
+            return $convertedMessage;
+        } catch (\Exception $e) {
+            throw $e;
+        }
+    }
+
+    /**
+     * @param array $dbalMessage
+     *
+     * @return MongodbMessage
+     */
+    protected function convertMessage(array $mongodbMessage)
+    {
+        $message = $this->context->createMessage();
```

Comment:
the createMessage method accept body, headers, and properties as arguments. pass them